core/log/profiler_hook.cpp

```diff
@@ -0,0 +1,378 @@
+    const stop::Criterion* criterion, const size_type& num_iterations,
+    const LinOp* residual, const LinOp* residual_norm, const LinOp* solution,
+    const uint8& stopping_id, const bool& set_finalized) const
+{
+    std::stringstream ss;
+    ss << "check(" << stringify_object(criterion) << ")";
+    this->begin_hook_(ss.str().c_str(), profile_event_category::criterion);
+}
+
+
+void ProfilerHook::on_criterion_check_completed(
+    const stop::Criterion* criterion, const size_type& num_iterations,
+    const LinOp* residual, const LinOp* residual_norm, const LinOp* solution,
+    const uint8& stopping_id, const bool& set_finalized,
+    const array<stopping_status>* status, const bool& one_changed,
+    const bool& all_converged) const
+{
+    this->on_criterion_check_completed(
+        criterion, num_iterations, residual, residual_norm, nullptr, solution,
+        stopping_id, set_finalized, status, one_changed, all_converged);
+}
+
+void ProfilerHook::on_criterion_check_completed(
+    const stop::Criterion* criterion, const size_type& num_iterations,
+    const LinOp* residual, const LinOp* residual_norm,
+    const LinOp* implicit_sq_resnorm, const LinOp* solution,
+    const uint8& stopping_id, const bool& set_finalized,
+    const array<stopping_status>* status, const bool& one_changed,
+    const bool& all_stopped) const
+{
+    std::stringstream ss;
+    ss << "check(" << stringify_object(criterion) << ")";
+    this->end_hook_(ss.str().c_str(), profile_event_category::criterion);
+}
+
+
+bool ProfilerHook::needs_propagation() const
+{
+    return true;
+}
+
+
+void ProfilerHook::set_object_name(const PolymorphicObject* obj,
+                                   std::string name)
+{
+    name_map_[obj] = name;
+}
+
+
+void ProfilerHook::set_synchronization(bool synchronize)
+{
+    synchronize_ = synchronize;
+}
+
+
+std::string ProfilerHook::stringify_object(const PolymorphicObject* obj) const
+{
+    if (!obj) {
+        return "nullptr";
+    }
+    auto it = name_map_.find(obj);
+    if (it != name_map_.end()) {
+        return it->second;
+    }
+    return name_demangling::get_dynamic_type(*obj);
+}
+
+
+ProfilerHook::ProfilerHook(hook_function begin, hook_function end)
+    : synchronize_{false}, begin_hook_{begin}, end_hook_{end}
+{}
+
+
+std::shared_ptr<ProfilerHook> ProfilerHook::create_tau(bool initialize)
+{
+    static std::mutex tau_mutex{};
+    static std::shared_ptr<ProfilerHook> tau_hook{};
+    std::lock_guard<std::mutex> guard{tau_mutex};
+    if (!tau_hook) {
+        if (initialize) {
+            init_tau();
+            tau_hook = std::shared_ptr<ProfilerHook>(
+                new ProfilerHook{begin_tau, end_tau}, [](ProfilerHook* ptr) {
+                    delete ptr;
+                    finalize_tau();
+                });
+        } else {
+            tau_hook = std::shared_ptr<ProfilerHook>{
+                new ProfilerHook{begin_tau, end_tau}};
+        }
+    }
```

Comment:
This would lead to shared data between different loggers created from `create_tau`. Perhaps just using a `static bool was_initialized` can be used to make sure that we only initialize and finalize it once:
```suggestion
    static bool was_initialized = false;
    std::shared_ptr<ProfilerHook> tau_hook{};
    std::lock_guard<std::mutex> guard{tau_mutex};
    if (was_initialized) {
        tau_hook =
            std::shared_ptr<ProfilerHook>{new ProfilerHook{begin_tau, end_tau}};
    } else {
        if (initialize) {
            was_initialized = true;
            init_tau();
            tau_hook = std::shared_ptr<ProfilerHook>(
                new ProfilerHook{begin_tau, end_tau}, [](ProfilerHook* ptr) {
                    delete ptr;
                    finalize_tau();
                });
        }
    }
```